File: lib/src/utils/path_utils.dart
String normalizePath(String path) => path.replaceAll('\\', '/');

String basename(String path) {
  final normalized = normalizePath(path);
  final slashIndex = normalized.lastIndexOf('/');
  return slashIndex == -1 ? normalized : normalized.substring(slashIndex + 1);
}

String dirname(String path) {
  final normalized = normalizePath(path);
  final slashIndex = normalized.lastIndexOf('/');
  if (slashIndex == -1) {
    return '.';
  }
  if (slashIndex == 0) {
    return '/';
  }
  return normalized.substring(0, slashIndex);
}

String joinPath(String directory, String name) {
  if (directory.isEmpty) return name;
  final normalized = normalizePath(directory);
  if (normalized.endsWith('/')) {
    return '$normalized$name';
  }
  return '$normalized/$name';
}

String relativePath(String rootPath, String fullPath) {
  final normalizedRoot = normalizePath(
    rootPath,
  ).replaceFirst(RegExp(r'/$'), '');
  final normalizedFull = normalizePath(fullPath);
  if (normalizedFull == normalizedRoot) {
    return '.';
  }

  final prefix = '$normalizedRoot/';
  if (normalizedFull.startsWith(prefix)) {
    return normalizedFull.substring(prefix.length);
  }

  return normalizedFull;
}

File: vscode-extension/src/utils/pathUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizePath = normalizePath;
exports.basename = basename;
exports.dirname = dirname;
exports.relativePath = relativePath;
function normalizePath(p) {
    return p.replace(/\\/g, '/');
}
function basename(p) {
    const normalized = normalizePath(p);
    const idx = normalized.lastIndexOf('/');
    return idx === -1 ? normalized : normalized.substring(idx + 1);
}
function dirname(p) {
    const normalized = normalizePath(p);
    const idx = normalized.lastIndexOf('/');
    if (idx === -1) {
        return '.';
    }
    if (idx === 0) {
        return '/';
    }
    return normalized.substring(0, idx);
}
function relativePath(rootPath, fullPath) {
    const normalizedRoot = normalizePath(rootPath).replace(/\/$/, '');
    const normalizedFull = normalizePath(fullPath);
    if (normalizedFull === normalizedRoot) {
        return '.';
    }
    const prefix = `${normalizedRoot}/`;
    if (normalizedFull.startsWith(prefix)) {
        return normalizedFull.substring(prefix.length);
    }
    return normalizedFull;
}
